lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
       end
     end
 
+    # Set required variables like @project_id, @vm_id, @vm_name and @zone.
+    def set_required_metadata_variables
+      set_project_id
+      set_vm_id
+      set_vm_name
+      set_location
+
+      # All metadata parameters must now be set.
+      return if @project_id && @zone && @vm_id
+      missing = []
+      missing << 'project_id' unless @project_id
+      missing << 'zone' unless @zone
+      missing << 'vm_id' unless @vm_id
+      fail Fluent::ConfigError, 'Unable to obtain metadata parameters: ' +
+        missing.join(' ')
+    end
+
+    def set_project_id
+      # 1. Check if the value is explicitly in the config thus already set.
+      return unless @project_id.nil?
+      # 2. Try to retrieve it by calling metadata servers directly.
+      @project_id = fetch_gce_metadata('project/project-id') if
+        @platform == Platform::GCE
+      # 3. Try to obtain it from the credentials.
+      @project_id ||= CredentialsInfo.project_id
+    rescue StandardError => e
+      @log.debug 'Failed to obtain project id: ', error: e
+    end
+
+    def set_vm_id
+      # 1. Check if the value is explicitly in the config thus already set.
+      return unless @vm_id.nil?
+      # 2. Check if the response from Metadata Agent includes this info.
+      @vm_id = @resource.labels['instance_id'] if
+        !@resource.nil? && @resource.labels.key?('instance_id')
+      # 3. Try to retrieve it by calling metadata servers directly.
+      @vm_id ||= fetch_gce_metadata('instance/id') if @platform == Platform::GCE
+      @vm_id ||= @ec2_metadata['instanceId'] if @platform == Platform::EC2
+    rescue StandardError => e
+      @log.debug 'Failed to obtain vm_id: ', error: e
+    end
+
+    def set_vm_name
+      # 1. Check if the value is explicitly in the config thus already set.
+      return unless @vm_name.nil?
+      # 2. Check if the response from Metadata Agent includes this info.
+      @vm_name ||= @resource.labels['instance_name'] if
+        !@resource.nil? && @resource.labels.key?('instance_name')
+      # 3. Detect it locally.
+      @vm_name ||= Socket.gethostname
+    rescue StandardError => e
+      @log.debug 'Failed to obtain vm name: ', error: e
+    end
+
+    def set_location
+      # 1. Check if the value is explicitly in the config thus already set.
+      return unless @zone.nil?
+      # 2. Check if the response from Metadata Agent includes this info.
+      unless @resource.nil?
+        @zone ||= @resource.labels['location'] if
+          @resource.labels.key?('location')
+        @zone ||= @resource.labels['zone'] if
+          @platform == Platform::GCE && @resource.labels.key?('zone')
+        @zone ||= @resource.labels['region'] if
```

Comment:
Changed. In fact I removed the entire section to extract the value from `@resource`. This is only meaningful after Metadata agent support is added. I left it here by mistake.